### PR TITLE
CT-347: Improve typescript support in BaseCoin

### PR DESCRIPTION
**Pull request by tylerlevine**
* Lift common coin interfaces into BaseCoin and use in subclasses
* Improve type support across project and enable `strictNullChecks` rule

### CODE REVIEW

File: modules/core/src/types/bitgo-utxo-lib.d.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    // bitgojs augmentation
+    getKey: (network: Network) => ECPair;
+  }
+
+  export class ECPair {
+    static makeRandom: ({ network: Network }) => ECPair;
+  }
+
+  export class KeyPair {
```
Comment (tylerlevine) [inline modules/core/src/types/bitgo-utxo-lib.d.ts]: Remove or finish defining

File: modules/core/src/v2/baseCoin.ts [hunk 2]
```diff
@@ -9,42 +9,125 @@ import * as Bluebird from 'bluebird';
...
 
 export interface KeyPair {
   pub?: string;
   prv: string;
 }
 
+export interface VerifyAddressOptions {
+  address: string;
+  addressType?: string;
+  chain?: number;
```
Comment (tylerlevine) [inline modules/core/src/v2/baseCoin.ts]: Move to child class in abstractUtxoCoin

File: modules/core/src/v2/baseCoin.ts [hunk 3]
```diff
@@ -9,42 +9,125 @@ import * as Bluebird from 'bluebird';
...
+
+export interface VerifyTransactionOptions {
+  txPrebuild: TransactionPrebuild;
+  txParams: TransactionParams;
+  wallet: Wallet;
+  verification?: VerificationOptions;
+  reqId?: RequestTracer;
+}
+
+export interface SupplementGenerateWalletOptions {}
```
Comment (tylerlevine) [inline modules/core/src/v2/baseCoin.ts]: Anything in here?

File: modules/core/src/v2/coins/abstractUtxoCoin.ts [hunk 4]
```diff
@@ -68,39 +77,30 @@ export interface ParsedSignatureScript {
...
-  txPrebuild: TxPrebuild;
+  txParams: TransactionParams;
+  txPrebuild: TransactionPrebuild;
   wallet: Wallet;
-  verification?: TxVerificationOptions;
+  verification?: VerificationOptions;
   reqId?: RequestTracer;
 }
 
 export interface ParsedTransaction {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/abstractUtxoCoin.ts]: Move up to baseCoin

File: modules/core/src/v2/coins/eos.ts [hunk 5]
```diff
@@ -310,19 +329,25 @@ export class Eos extends BaseCoin {
    * @param address - the address to verify
    * @param rootAddress - the wallet's root address
    */
-  verifyAddress({ address, rootAddress }: { address: string; rootAddress: string }): void {
-    if (!this.isValidAddress(address)) {
+  verifyAddress({ address, rootAddress }: VerifyAddressOptions): boolean {
+    if (!rootAddress || !this.isValidAddress(address)) {
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/eos.ts]: separate conditions

File: modules/core/src/v2/coins/xrp.ts [hunk 6]
```diff
@@ -561,18 +567,21 @@ export class Xrp extends BaseCoin {
...
-      const queryDetails = querystring.parse(destinationDetails.query);
       const destinationAddress = destinationDetails.pathname;
-      let destinationTag = undefined;
 
-      if (Array.isArray(queryDetails.dt)) {
-        // if queryDetails.dt is an array, that means dt was given multiple times, which is not valid
-        throw new InvalidAddressError(`destination tag can appear at most once, but ${queryDetails.dt.length} destination tags were found`);
-      }
+      // parse destination tag from query
+      let destinationTag: number | undefined;
```
Comment (tylerlevine) [inline modules/core/src/v2/coins/xrp.ts]: take a pass at improving this to be a bit flatter

File: modules/core/src/v2/markets.ts [hunk 7]
```diff
@@ -66,7 +66,7 @@ export class Markets {
     return co(function*() {
       validateParams(params, ['currencyName'], []);
 
-      const days = !isNaN(parseInt(params.days, 10)) ? parseInt(params.days, 10) : 90;
+      const days = params.days && !isNaN(parseInt(params.days, 10)) ? parseInt(params.days, 10) : 90;
```
Comment (tylerlevine) [inline modules/core/src/v2/markets.ts]: check for !undefined instead of truthy

File: modules/core/src/v2/trading/settlements.ts [hunk 8]
```diff
@@ -59,18 +64,18 @@ export class Settlements {
    * @param accountId ID of the trading account that the affirmation belongs to
    * @param callback
    */
-  get({ id, accountId }, callback?: NodeCallback<Settlement>): Bluebird<Settlement> {
+  get({ id, accountId }: GetOptions, callback?: NodeCallback<Settlement>): Bluebird<Settlement> {
     const self = this;
     return co(function* get() {
-      if (!accountId && !self.account) {
+      const account = accountId || (self.account && self.account.id);
+      if (!account) {
```
Comment (tylerlevine) [inline modules/core/src/v2/trading/settlements.ts]: check for undefined

File: modules/core/src/transactionBuilder.ts [hunk 9]
```diff
@@ -360,7 +379,11 @@ exports.createTransaction = function(params) {
 
   // Iterate unspents, sum the inputs, and save _inputs with the total
   // input amount and final list of inputs to use with the transaction.
-  let feeSingleKeyUnspentsUsed = [];
+  let feeSingleKeyUnspentsUsed: {
+    value: number;
+    tx_hash: Buffer;
+    tx_output_n: number;
+  }[] = [];
```
Comment (dannybitgo) [inline modules/core/src/transactionBuilder.ts]: this is repeated from lines 356-360. doesn't bother me too much, but might be worth defining the type somewhere up top instead

File: modules/core/src/v2/baseCoin.ts [hunk 10]
```diff
@@ -330,22 +393,21 @@ export abstract class BaseCoin {
...
-          return Promise.resolve(userHDNode);
+          return bitcoin.HDNode.fromBase58(userKey);
         } catch (e) {
           throw new Error('Failed to decrypt user key with passcode - try again!');
         }
-      };
+      }
 
-      const key = yield validatePassphraseKey(userKey, passphrase);
+      const key: bitcoin.HDNode = validatePassphraseKey(userKey, passphrase);
```
Comment (dannybitgo) [inline modules/core/src/v2/baseCoin.ts]: no more ```yield``` ?

File: modules/core/src/v2/coins/xrp.ts (hunk 6 above)
Comment (tylerlevine) [reply modules/core/src/v2/coins/xrp.ts]: deferring to future pr

File: modules/core/src/v2/markets.ts (hunk 7 above)
Comment (tylerlevine) [reply modules/core/src/v2/markets.ts]: deferring to future pr

File: modules/core/src/v2/baseCoin.ts (hunk 10 above)
Comment (tylerlevine) [reply modules/core/src/v2/baseCoin.ts]: nope. `validatePassphraseKey` is a function which is internal to this outer function, and it never needed to be async in the first place.

File: modules/core/src/transactionBuilder.ts (hunk 9 above)
Comment (tylerlevine) [reply modules/core/src/transactionBuilder.ts]: done

**Pull request merged**